In [1]:
# importamos las librerías que necesitamos

# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np
from src import soporte_datasets as sp_datasets

# Visualización
# -----------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns

# Imputación de nulos usando métodos avanzados estadísticos
# -----------------------------------------------------------------------
from sklearn.impute import SimpleImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.impute import KNNImputer


# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

# Gestión de los warnings
# -----------------------------------------------------------------------
import warnings
warnings.filterwarnings("ignore")

In [3]:
df_final = pd.read_csv('files/customer_flight_loyalty.csv') 
df_final.head(2)

,loyalty_number,year,month,flights_booked,flights_with_companions,total_flights,distance,points_accumulated,points_redeemed,dollar_cost_points_redeemed,province,city,postal_code,gender,education,marital_status,loyalty_card,clv,enrollment_type,enrollment_year,enrollment_month,cancellation_year,cancellation_month,salary
0,100018,2017,1,3,0,3,1521,152.0,0,0,Alberta,Edmonton,T9G 1W3,Female,Bachelor,Married,Aurora,7919.20,Standard,2016,8,False,False,92552.0
1,100102,2017,1,10,4,14,2030,203.0,0,0,Ontario,Toronto,M1R 4K3,Male,College,Single,Nova,2887.74,Standard,2013,3,False,False,66937.5


In [5]:
df_final.columns

Index(['loyalty_number', 'year', 'month', 'flights_booked',
       'flights_with_companions', 'total_flights', 'distance',
       'points_accumulated', 'points_redeemed', 'dollar_cost_points_redeemed',
       'province', 'city', 'postal_code', 'gender', 'education',
       'marital_status', 'loyalty_card', 'clv', 'enrollment_type',
       'enrollment_year', 'enrollment_month', 'cancellation_year',
       'cancellation_month', 'salary'],
      dtype='str')

In [11]:
#Análisis de variables numéricas

df_final.describe().T

,count,mean,std,min,25%,50%,75%,max
loyalty_number,403760.0,549875.383713,258961.514684,100018.00,326699.00,550598.00,772152.00,999986.00
year,403760.0,2017.500352,0.500000,2017.00,2017.00,2018.00,2018.00,2018.00
month,403760.0,6.501335,3.451982,1.00,4.00,7.00,10.00,12.00
flights_booked,403760.0,4.134050,5.230064,0.00,0.00,1.00,8.00,21.00
flights_with_companions,403760.0,1.036569,2.080472,0.00,0.00,0.00,1.00,11.00
total_flights,403760.0,5.170619,6.526858,0.00,0.00,1.00,10.00,32.00
distance,403760.0,1214.460979,1434.098521,0.00,0.00,525.00,2342.00,6293.00
points_accumulated,403760.0,124.263761,146.696179,0.00,0.00,53.00,240.00,676.50
points_redeemed,403760.0,30.838587,125.758002,0.00,0.00,0.00,0.00,876.00
dollar_cost_points_redeemed,403760.0,2.495973,10.172033,0.00,0.00,0.00,0.00,71.00


Podemos interpretar lo siguiente:

    - flights_booked: El 75% de los datos llega a 8, pero el máximo es 21, podemos interpretar presencia de outliers.
    - flights_with_companions: La media está en 1m y el máximo en 11, muy disperso.
    - total_flights: El 75% de los datos llega a 10, pero el máximo es 32, podemos interpretar presencia de outliers.
    - distance: La media es más del doble que la mediana, y el valor máximo está muy alejado del 75% de los datos, presencia de outliers.
    - points_accumulated: La media está en 124 frente a 53 de la mediana (el doble) teniendo un máximo de 676,50, muy disperso.
    - points_redeemed: La media de puntos gastados son 30 mensuales, la mediana se sitúa en 0, teniendo un máximo de 876, datos muy dispersos.
    - dollar_cost_points_redeemed: La media del valor en dólares es 10, la mediana se sitúa en 0, teniendo un máximo de 71$, datos muy dispersos.
    - clv: La media es mayor que la mediana (7990 vs 5776), el 75% de los clientes aportan 8937$ pero tiene cliente(s) que aportan hasta 83325, muy disperso.
    - salary: El salario máximo (407228) es altísimo comparado con la media (76222) o mediana /66937.

    

In [10]:
#¿Cómo se distribuye la cantidad de vuelos reservados por mes durante el año?

df_final[["total_flights", "flights_booked"]].agg("mean")

total_flights     5.170619
flights_booked    4.134050
dtype: float64

In [ ]:
#Análisis correlación entre variables numéricas

for col in df_final.columns: #pendiente if para que coja las numericas. dtype
    limite, cantidad = detectar_outliers_iqr(df_final, col)
    porcentaje = (cantidad / len(df_final)) * 100
    print(f"{col}: Límite superior {limite:.2f} | Cantidad: {cantidad} ({porcentaje:.2f}%)")


plt.figure(figsize=(10, 6))
correlacion = df_final[cols_num].corr()
sns.heatmap(correlacion, annot=True, cmap='coolwarm', fmt=".2f")
plt.title("Matriz de Correlación entre Variables Numéricas")
plt.show()

NameError: name 'cols_num' is not defined

<Figure size 1000x600 with 0 Axes>